# Assignment 2
### Normal Distribution, Sampling & Confidence

### Dataset

Breakout is a fun arcade game by *Atari Inc.* from 1976 where you control a small paddle that can move sideways at the bottom of the screen. The goal is to bounce a ball back upwards against placed bricks. Each brick that is broken awards the player with some points. When all bricks are cleared, the player advances to the next level. Whenever the ball is not returned, the player loses a life and can try again in the same level. When all lives are lost, the game is over.

Two friends play this game simultaneously, and define winning as having a higher score than the other person. All the games these two friends ever played together are collected in this dataset. 

Let's start by loading the toolboxes we need and uploading the dataset. 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import norm

sns.set_style("whitegrid")

In [20]:
### needed/convenient only if you're working in Google colab (not for local Jupyter notebook):
# from google.colab import files
# uploaded = files.upload()

In [5]:
breakout_score = pd.read_csv("Breakout.csv")
breakout_score.head()

score_1 wins_1  score_2 wins_2
0    16500      L    19400      W
1    18600      L    21100      W
2    27100      W    25400      L
3    19800      L    26800      W
4    12600      L    16600      W

In this workgroup, we will focus only on the scores of player 1 which you can find in the variable `score_1`. Each row represents the score of one game. 

### Exercise 1

How often did player 1 play the game? <div style='text-align: right;'>**1 point**</div>

In [1]:
# your code/answer here

#### Normal distribution

Before we are going to analyze the data, it is important to see how the data is distributed, or more importantly, if it is normally distributed or not.

### Exercise 2

Plot the distributions of the scores of this player, and describe the distribution. Do you think the data are normally distributed? (hint: use a histplot) <div style='text-align: right;'>**2 points**</div>

In [2]:
# your code/answer here

#### Normal probability

Once we decide that a variable is approximately normal, we can use our knowledge about the normal distribution to answer all sorts of questions about that variable related to probability. Take, for example, the question "*what is the probability that this player scores 20000 or less in a particular game?*"

To answer this question, we compute the z-score, which tells us something about how far a data point is from the mean. Then we use a z-table to convert this z-score to a probability. 

In Python, this can be done in one step with the function `stats.norm.cdf()`:

In [9]:
mean = breakout_score['score_1'].mean()
std = breakout_score['score_1'].std()
score = 20000
norm.cdf(score, mean, std)

0.4075083828254088

This means that this player scores less than 20000 points with a probability of 0.408. 

We can compare this probability to the empirical probability. We simply need to determine how many observations fall below 20000 then divide this number by the total sample size. 

In [10]:
sum(breakout_score['score_1'] <= score) / len(breakout_score)

0.411


Although the probabilities are not exactly the same, they are reasonably close. The closer the distribution is to being normal, the more accurate the theoretical probabilities will be.

### Exercise 3

The average score of all players online is 21500. What is the probability that, in a particular game, player 1 scores higher than this online average? Use the normal approximation to compute this probability. When printing out the result from any calculation, include some verbal description of what the numbers mean.  <div style='text-align: right;'>**2 points**</div>

In [3]:
# your code/answer here

#### Sampling

In this lab we have access to the entire population, but this is rarely the case in real life. Gathering information on an entire population is often extremely costly or impossible. Because of this, we often take a sample of the population and use that to understand the properties of the population.

We can use the following command to sample 30 observations from the population:

In [23]:
samp = breakout_score['score_1'].sample(30)

### Exercise 4

Have a look at the distribution of the sampled data by making another displot or histplot. Describe how it differs from the population distribution. And how does that change when you draw a larger sample of 100 scores? <div style='text-align: right;'>**3 points**</div>

In [4]:
# your code/answer here

### Exercise 5

Let's recompute our estimate for the probability that in a particular game player 1 scores higher than the online average. 

1. Does the probability differ from the probability observed in the population? 
2. Which sample estimate (30 or 100 participants) is closer to the population estimate?
3. Do your observations from 1. and 2. above hold for every random sample (repeat the sampling multiple times)?<div style='text-align: right;'>**4 points**</div>

In [5]:
# your code/answer here

### Exercise 6

Every time we take another random sample, we get a different distribution and a different estimate.   It's useful to get a sense of just how much variability we should expect when estimating the population proportion this way. The distribution of sample probabilities, called the sampling distribution, can help us understand this variability. We can build up the sampling distribution for the population proportion by repeating the above steps many times. 

1. Create 5000 samples of size 30 and size 100 of the player's scores (hint: create empty numpy arrays and fill those with values in your for-loop). 
2. For each sample, compute the probability under the normal distribution that this player outperforms the average player. 
3. Save the values in each sample and show them in a histogram (you can use the histplot again). Make one histogram per sample size and make sure the x-axes are equal! 
4. How do the center and the spread of your sample distribution change when your sample size increases? <div style='text-align: right;'>**6 points**</div>

In [6]:
# your code/answer here

In [7]:
# your code/answer here

#### Confidence intervals

Let's say we want to know what the average performance of our player is. If we have access to data from all games (population), it is straightforward to answer questions like this. When we only have sample data, we can approximate the mean of our population using our sample mean. That serves as a good point estimate but it would be useful to also communicate how uncertain we are of that estimate. This can be captured by using a confidence interval.

In the online lecture, you have seen how to calculate a confidence interval for probabilities. We can use a similar method to compute a confidence interval for numerical data. We can calculate a 95% confidence interval for a sample mean by adding and subtracting 1.96 standard errors to the point estimate (this is the z*–value). By computing confidence intervals, we can make inferences about what the population means. For numerical data, we can compute the standard error using the formula: SE = $\frac{\sigma}{\sqrt n}$


### Exercise 7

1. Write a function that takes a sample dataset and a z*–value as input and outputs the lower and the upper bound of the confidence interval.
2. Draw a sample of 30 scores and print the confidence interval. Does the confidence interval contain the population mean?
3. Now recompute the confidence interval for a sample of 200 scores. How does the interval change? Why is that?
4. Now change the confidence level to 99% (z*=2.58) for the 200 samples. How does that change your interval?
5. Draw 5000 samples of 50 scores and calculate the percentage of how often the 95% confidence interval contains the population mean. Is this number similar to what you would expect?<div style='text-align: right;'>**11 points**</div>

In [41]:
# your code/answer here

In [8]:
# your code/answer here

In [9]:
# your code/answer here

In [10]:
# your code/answer here

In [11]:
# your code/answer here

**Total number of points**: 1 + 2 + 2 + 3 + 4 + 6 + 11 = 29